Tests randomly fail due to KeyboardInterrupt

# Bug Description

Tests for version 2.1.0 fail often with errors similar to:

```
==================================== ERRORS ====================================
_____________ ERROR at teardown of test_03_publish_helper_qos0_v5 ______________
    def fin():
        stop_process(proc)
        if proc.returncode != expected_returncode:
>           raise RuntimeError(f"Client {name} exited with code {proc.returncode}, expected {expected_returncode}")
E           RuntimeError: Client 03-publish-helper-qos0-v5.py exited with code -2, expected 0
tests/lib/conftest.py:75: RuntimeError
--------------------------- Captured stderr teardown ---------------------------
Traceback (most recent call last):
  File "/build/source/tests/lib/clients/03-publish-helper-qos0-v5.py", line 7, in <module>
    paho.mqtt.publish.single(
  File "/nix/store/sg4pgfbqcb5apkrsiq1npxvnwd4in57j-python3.12-paho-mqtt-2.1.0/lib/python3.12/site-packages/paho/mqtt/publish.py", line 305, in single
    multiple([msg], hostname, port, client_id, keepalive, will, auth, tls,
  File "/nix/store/sg4pgfbqcb5apkrsiq1npxvnwd4in57j-python3.12-paho-mqtt-2.1.0/lib/python3.12/site-packages/paho/mqtt/publish.py", line 228, in multiple
    client.loop_forever()
  File "/nix/store/sg4pgfbqcb5apkrsiq1npxvnwd4in57j-python3.12-paho-mqtt-2.1.0/lib/python3.12/site-packages/paho/mqtt/client.py", line 2297, in loop_forever
    rc = self._loop(timeout)
         ^^^^^^^^^^^^^^^^^^^
  File "/nix/store/sg4pgfbqcb5apkrsiq1npxvnwd4in57j-python3.12-paho-mqtt-2.1.0/lib/python3.12/site-packages/paho/mqtt/client.py", line 1703, in _loop
    rc = self.loop_write()
         ^^^^^^^^^^^^^^^^^
  File "/nix/store/sg4pgfbqcb5apkrsiq1npxvnwd4in57j-python3.12-paho-mqtt-2.1.0/lib/python3.12/site-packages/paho/mqtt/client.py", line 2121, in loop_write
    rc = self._packet_write()
         ^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/sg4pgfbqcb5apkrsiq1npxvnwd4in57j-python3.12-paho-mqtt-2.1.0/lib/python3.12/site-packages/paho/mqtt/client.py", line 3230, in _packet_write
    self._sock_close()
  File "/nix/store/sg4pgfbqcb5apkrsiq1npxvnwd4in57j-python3.12-paho-mqtt-2.1.0/lib/python3.12/site-packages/paho/mqtt/client.py", line 1135, in _sock_close
    sock.close()
  File "/nix/store/rdq5zjrqdf9515q0kxps2dklhq7ybnwh-python3-3.12.10/lib/python3.12/socket.py", line 505, in close
    self._real_close()
  File "/nix/store/rdq5zjrqdf9515q0kxps2dklhq7ybnwh-python3-3.12.10/lib/python3.12/socket.py", line 499, in _real_close
    _ss.close(self)
KeyboardInterrupt
=========================== short test summary info ============================
XFAIL tests/lib/test_03_publish_fill_inflight.py::test_03_publish_fill_inflight
============= 162 passed, 1 xfailed, 1 error in 132.09s (0:02:12) ==============
```

# Reproduction

1. Run the test suite, using Python 3.12. Here done as part of the [nixpkgs](https://github.com/NixOS/nixpkgs) package.

The issue is not consistently reproducible.

# Environment

* Python version: 3.12
* Library version: 2.1.0
* Operating system (including version): Linux (NixOS 25.05 unstable)
* MQTT server (name, version, configuration, hosting details): N/A

--- a/tests/lib/conftest.py
+++ b/tests/lib/conftest.py
@@ -2,6 +2,7 @@ import os
 import signal
 import subprocess
 import sys
+import time
 
 import pytest
 
@@ -45,7 +46,29 @@ def alpn_ssl_server_socket(monkeypatch, ssl_certs_path):
     yield from _yield_server(monkeypatch, create_server_socket_ssl(path=ssl_certs_path, alpn_protocols=["paho-test-protocol"]))
 
 
-def stop_process(proc: subprocess.Popen) -> None:
+def terminate_process(proc: subprocess.Popen) -> None:
+    proc.terminate()
+    try:
+        # At least on Unix, terminate() isn't an unconditional kill, process
+        # could ignore/handle it.
+        proc.wait(1)
+    except subprocess.TimeoutExpired:
+        # So use kill which is unstoppable on Unix
+        proc.kill()
+        proc.wait(1)  # If here we timeout, there is nothing more to do.
+
+
+def stop_process(proc: subprocess.Popen, ready_file) -> None:
+    deadline = time.monotonic() + 5
+    while proc.poll() is None and not ready_file.exists():
+        if time.monotonic() >= deadline:
+            terminate_process(proc)
+            raise RuntimeError("Client did not become ready to stop")
+        time.sleep(0.01)
+
+    if proc.poll() is not None:
+        return
+
     if sys.platform == "win32":
         proc.send_signal(signal.CTRL_C_EVENT)
     else:
@@ -53,25 +76,19 @@ def stop_process(proc: subprocess.Popen) -> None:
     try:
         proc.wait(5)
     except subprocess.TimeoutExpired:
-        proc.terminate()
-        try:
-            # At least on Unix, terminate() isn't an unconditional kill, process
-            # could ignore/handle it.
-            proc.wait(1)
-        except subprocess.TimeoutExpired:
-            # So use kill which is unstoppable on Unix
-            proc.kill()
-            proc.wait(1)  # If here we timeout, there is nothing more to do.
+        terminate_process(proc)
 
 
 @pytest.fixture()
-def start_client(request: pytest.FixtureRequest, ssl_certs_path):
+def start_client(request: pytest.FixtureRequest, ssl_certs_path, tmp_path):
     def starter(name: str, expected_returncode: int = 0) -> None:
         client_path = clients_path / name
         if not client_path.exists():
             raise FileNotFoundError(client_path)
+        ready_file = tmp_path / f"{name}.ready"
         env = dict(
             os.environ,
+            PAHO_TEST_READY_FILE=str(ready_file),
             PAHO_SSL_PATH=str(ssl_certs_path),
             PYTHONPATH=f"{tests_path}{os.pathsep}{os.environ.get('PYTHONPATH', '')}",
         )
@@ -82,7 +99,7 @@ def start_client(request: pytest.FixtureRequest, ssl_certs_path):
         ], env=env)
 
         def fin():
-            stop_process(proc)
+            stop_process(proc, ready_file)
             # If return code is None, process had not stopped when a method was last called on proc.
             # https://docs.python.org/library/subprocess.html#subprocess.Popen.returncode
             if proc.returncode != expected_returncode:

--- a/tests/paho_test.py
+++ b/tests/paho_test.py
@@ -403,6 +403,13 @@ def pack_remaining_length(remaining_length):
             return s
 
 
+def signal_client_ready():
+    ready_file = os.environ.get("PAHO_TEST_READY_FILE")
+    if ready_file is not None:
+        with open(ready_file, "w"):
+            pass
+
+
 def loop_until_keyboard_interrupt(mqttc):
     """
     Call loop() in a loop until KeyboardInterrupt is received.
@@ -413,6 +420,7 @@ def loop_until_keyboard_interrupt(mqttc):
     and stop the client gracefully.
     """
     try:
+        signal_client_ready()
         while True:
             mqttc.loop()
     except KeyboardInterrupt:
@@ -431,6 +439,7 @@ def wait_for_keyboard_interrupt():
     """
     yield  # If we get a KeyboardInterrupt during the block, it's too soon!
     try:
+        signal_client_ready()
         while True:
             time.sleep(0.1)
     except KeyboardInterrupt:

--- a/tests/test_client.py
+++ b/tests/test_client.py
@@ -864,18 +864,18 @@ class TestCompatibility:
             disconnect_packet = paho_test.gen_disconnect()
             fake_broker.expect_packet("disconnect", disconnect_packet)
 
-            assert callback_called == [
-                "on_connect",
-                "on_subscribe",
-                "on_publish",
-                "on_message",
-                "on_unsubscribe",
-                "on_disconnect",
-            ]
-
         finally:
             mqttc.disconnect()
             mqttc.loop_stop()
+
+        assert callback_called == [
+            "on_connect",
+            "on_subscribe",
+            "on_publish",
+            "on_message",
+            "on_unsubscribe",
+            "on_disconnect",
+        ]
 
         packet_in = fake_broker.receive_packet(1)
         assert not packet_in  # Check connection is closed
@@ -1004,18 +1004,18 @@ class TestCompatibility:
             disconnect_packet = paho_test.gen_disconnect()
             fake_broker.expect_packet("disconnect", disconnect_packet)
 
-            assert callback_called == [
-                "on_connect",
-                "on_subscribe",
-                "on_publish",
-                "on_message",
-                "on_unsubscribe",
-                "on_disconnect",
-            ]
-
         finally:
             mqttc.disconnect()
             mqttc.loop_stop()
+
+        assert callback_called == [
+            "on_connect",
+            "on_subscribe",
+            "on_publish",
+            "on_message",
+            "on_unsubscribe",
+            "on_disconnect",
+        ]
 
         packet_in = fake_broker.receive_packet(1)
         assert not packet_in  # Check connection is closed

--- a/tests/testsupport/broker.py
+++ b/tests/testsupport/broker.py
@@ -59,8 +59,10 @@ class FakeBroker:
         if self._conn is None:
             raise ValueError('Connection is not open')
 
-        packet_in = self._conn.recv(num_bytes)
-        return packet_in
+        try:
+            return self._conn.recv(num_bytes)
+        except ConnectionResetError:
+            return b""
 
     def send_packet(self, packet_out):
         if self._conn is None:

